src/main/java/io/numaproj/numaflow/mapper/Server.java

```diff
@@ -56,43 +63,28 @@ public Server(Mapper mapper, GRPCConfig grpcConfig) {
      * @throws Exception if the server fails to start
      */
     public void start() throws Exception {
-
-        if (!grpcConfig.isLocal()) {
+        if (!this.grpcConfig.isLocal()) {
             GrpcServerUtils.writeServerInfo(
-                    serverInfoAccessor,
-                    grpcConfig.getSocketPath(),
-                    grpcConfig.getInfoFilePath(),
+                    this.serverInfoAccessor,
+                    this.grpcConfig.getSocketPath(),
+                    this.grpcConfig.getInfoFilePath(),
                     ContainerType.MAPPER,
                     Collections.singletonMap(Constants.MAP_MODE_KEY, Constants.MAP_MODE));
         }
 
-        if (this.server == null) {
-            this.server = this.grpcServerHelper.createServer(
-                    grpcConfig.getSocketPath(),
-                    grpcConfig.getMaxMessageSize(),
-                    grpcConfig.isLocal(),
-                    grpcConfig.getPort(),
-                    this.service);
-        }
-
-        server.start();
+        this.server.start();
 
         log.info(
                 "server started, listening on {}",
-                grpcConfig.isLocal() ?
-                        "localhost:" + grpcConfig.getPort() : grpcConfig.getSocketPath());
+                this.grpcConfig.isLocal() ?
+                        "localhost:" + this.grpcConfig.getPort() : this.grpcConfig.getSocketPath());
 
         // register shutdown hook to gracefully shut down the server
         Runtime.getRuntime().addShutdownHook(new Thread(() -> {
             // Use stderr here since the logger may have been reset by its JVM shutdown hook.
             System.err.println("*** shutting down gRPC server since JVM is shutting down");
-            if (server != null && server.isTerminated()) {
-                return;
-            }
             try {
-                Server.this.stop();
-                log.info("gracefully shutting down event loop groups");
-                this.grpcServerHelper.gracefullyShutdownEventLoopGroups();
+                this.stop();
                 // FIXME - this is a workaround to immediately terminate the JVM process
                 // The correct way to do this is to stop all the actors and wait for them to terminate
                 System.exit(0);
```

Comment:
Anyway to get rid of hard exit?

Reply:
Thanks, @RohanAshar . Yes, I plan to fix it very soon. I was not able to fix it this week so I decided to create an [issue](https://github.com/numaproj/numaflow-java/issues/155) and fix it ASAP next week.